File: FrontEnd/routes/fall_trained.js
const express = require('express');
const router = express.Router();
const fs = require('fs');
const cv = require('opencv4nodejs');
const tf = require('@tensorflow/tfjs');
const { loadGraphModel } = require('@tensorflow/tfjs-converter');
const { createCanvas, Image } = require('canvas');

let fall_detection_result = false;

const BODY_PARTS = {"Nose": 0, "Neck": 1, "RShoulder": 2, "RElbow": 3, "RWrist": 4,
              "LShoulder": 5, "LElbow": 6, "LWrist": 7, "RHip": 8, "RKnee": 9,
              "RAnkle": 10, "LHip": 11, "LKnee": 12, "LAnkle": 13, "REye": 14,
              "LEye": 15, "REar": 16, "LEar": 17};

const POSE_PAIRS = [["Neck", "RShoulder"], ["Neck", "LShoulder"], ["RShoulder", "RElbow"],
              ["RElbow", "RWrist"], ["LShoulder", "LElbow"], ["LElbow", "LWrist"],
              ["Neck", "RHip"], ["RHip", "RKnee"], ["RKnee", "RAnkle"], ["Neck", "LHip"],   
              ["LHip", "LKnee"], ["LKnee", "LAnkle"], ["Neck", "Nose"], ["Nose", "REye"],
              ["REye", "REar"], ["Nose", "LEye"], ["LEye", "LEar"]];

const EDGES = {};
              EDGES[BODY_PARTS["Nose"], BODY_PARTS["Neck"]] = 'm';
              EDGES[BODY_PARTS["Neck"], BODY_PARTS["RShoulder"]] = 'm';
              EDGES[BODY_PARTS["Neck"], BODY_PARTS["LShoulder"]] = 'c';
              EDGES[BODY_PARTS["RShoulder"], BODY_PARTS["RElbow"]] = 'm';
              EDGES[BODY_PARTS["RElbow"], BODY_PARTS["RWrist"]] = 'm';
              EDGES[BODY_PARTS["LShoulder"], BODY_PARTS["LElbow"]] = 'c';
              EDGES[BODY_PARTS["LElbow"], BODY_PARTS["LWrist"]] = 'c';
              EDGES[BODY_PARTS["Neck"], BODY_PARTS["RHip"]] = 'y';
              EDGES[BODY_PARTS["RHip"], BODY_PARTS["RKnee"]] = 'm';
              EDGES[BODY_PARTS["RKnee"], BODY_PARTS["RAnkle"]] = 'm';
              EDGES[BODY_PARTS["Neck"], BODY_PARTS["LHip"]] = 'y';
              EDGES[BODY_PARTS["LHip"], BODY_PARTS["LKnee"]] = 'c';
              EDGES[BODY_PARTS["LKnee"], BODY_PARTS["LAnkle"]] = 'c';
              EDGES[BODY_PARTS["Nose"], BODY_PARTS["REye"]] = 'm';
              EDGES[BODY_PARTS["REye"], BODY_PARTS["REar"]] = 'm';
              EDGES[BODY_PARTS["Nose"], BODY_PARTS["LEye"]] = 'c';
              EDGES[BODY_PARTS["LEye"], BODY_PARTS["LEar"]] = 'c';
              

const net = cv.readNetFromTensorflow('./graph_opt.pb');

const inWidth = 368;
const inHeight = 368;
const thr = 0.2;

const fall_model_path = "/app/export/fall-detection-model.h5";
let fall_model;
async function loadFallModel() {
    try {
        fall_model = await tf.loadLayersModel('file://' + fall_model_path);
    } catch (error) {
        console.error("Error loading fall detection model:", error);
    }
}
loadFallModel();

const interpreter = new tf.lite.Interpreter({ modelPath: "/app/3.tflite" });
interpreter.allocateTensors();

function runMovenet(image) {
    const canvas = createCanvas(192, 192);
    const ctx = canvas.getContext('2d');
    const img = new Image();
    img.src = image;
    ctx.drawImage(img, 0, 0, 192, 192);
    const input = tf.browser.fromPixels(canvas);
    const resizedInput = tf.image.resizeBilinear(input, [192, 192]);
    const expandedInput = resizedInput.expandDims(0);
    const inputDetails = interpreter.getInputDetails();
    const outputDetails = interpreter.getOutputDetails();
    interpreter.setTensor(inputDetails[0].index, expandedInput);
    interpreter.invoke();
    const keypointsWithScores = interpreter.getTensor(outputDetails[0].index);
    const keypoints = keypointsWithScores.slice([0, 0, 0], [1, -1, 2]);
    return keypoints.arraySync()[0];
}

function drawKeypoints(frame, keypoints) {
    const shaped = keypoints.map(kp => [kp[0] * frame.cols, kp[1] * frame.rows]);
    shaped.forEach(([kx, ky]) => {
        frame.drawCircle(new cv.Point(kx, ky), 4, new cv.Vec(0, 255, 0), -1);
        frame.putText(`(${parseInt(kx)}, ${parseInt(ky)})`, new cv.Point(kx + 10, ky + 10), cv.FONT_HERSHEY_SIMPLEX, 0.5, new cv.Vec(255, 255, 255), 1);
    });
}

function drawConnections(frame, keypoints, edges) {
    const shaped = keypoints.map(kp => [kp[0] * frame.cols, kp[1] * frame.rows]);
    edges.forEach(([p1, p2]) => {
        const [x1, y1] = shaped[p1];
        const [x2, y2] = shaped[p2];
        frame.drawLine(new cv.Point(x1, y1), new cv.Point(x2, y2), new cv.Vec(0, 0, 255), 2);
    });
}

function poseEstimation(frame) {
    const blob = cv.blobFromImage(frame, 1.0, new cv.Size(inWidth, inHeight), new cv.Vec(127.5, 127.5, 127.5), true, false);
    net.setInput(blob);
    let out = net.forward();
    out = out.slice([0, 0, 0], [1, out.sizes[1], out.sizes[2]]);
    const points = [];
    const confidences = [];
    for (let i = 0; i < BODY_PARTS.length; i++) {
        const heatMap = out.at(0, i);
        const [_, conf, __, point] = cv.minMaxLoc(heatMap);
        const x = (frame.cols * point.x) / out.sizes[3];
        const y = (frame.rows * point.y) / out.sizes[2];
        points.push(conf > thr ? new cv.Point(x, y) : null);
        confidences.push(conf);
    }
    for (const pair of POSE_PAIRS) {
        const [partFrom, partTo] = pair;
        const idFrom = BODY_PARTS[partFrom];
        const idTo = BODY_PARTS[partTo];
        if (points[idFrom] && points[idTo]) {
            frame.drawLine(points[idFrom], points[idTo], new cv.Vec(0, 255, 0), 3);
            frame.drawEllipse(points[idFrom], new cv.Size(3, 3), 0, 0, 360, new cv.Vec(0, 0, 255), cv.FILLED);
            frame.drawEllipse(points[idTo], new cv.Size(3, 3), 0, 0, 360, new cv.Vec(0, 0, 255), cv.FILLED);
        }
    }
    return [frame, points, confidences];
}

async function predictImage(imagePath) {
    try {
        if (!fall_model) {
            console.error("Fall detection model not loaded");
            return false;
        }
        
        const image = cv.imread(imagePath);
        const keypoints = runMovenet(image);
        const flattenedKeypoints = keypoints.flat();
        const tensorInput = tf.tensor2d(flattenedKeypoints, [1, flattenedKeypoints.length]);
        const prediction = fall_model.predict(tensorInput).dataSync()[0];
        return prediction > 0.5;
    } catch (error) {
        console.error("Error predicting:", error);
        return false;
    }
}

router.get('/', (req, res) => {
    res.sendFile(__dirname + '/static/index.html');
});

router.use(express.static('static'));

router.post('/upload', async (req, res) => {
    try {
        const content = req.files.content;
        const imagePath = '/tmp/in.png';
        content.mv(imagePath, async (err) => {
            if (err) {
                console.error("Error saving image:", err);
                res.status(500).send("Error saving image");
            } else {
                const img = cv.imread(imagePath);
                const fallDetected = await predictImage(imagePath);
                fall_detection_result = fallDetected;
                console.log(fallDetected);
                cv.imwrite('/tmp/out.png', img);
                res.send("ok");
            }
        });
    } catch (error) {
        console.error("Error:", error);
        res.status(500).send("Error: " + error.message);
    }
});

router.get('/get_fall_detection_result', (req, res) => {
    res.send(fall_detection_result.toString());
});

router.get('/get', (req, res) => {
    res.sendFile(__dirname + '/tmp/out.png');
});

module.exports = router;

File: FrontEnd/routes/fall_trained.py
import cv2
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf
from tensorflow.keras.models import load_model
from tensorflow.keras.preprocessing import image
from tensorflow.keras.preprocessing.image import load_img

from flask import Flask, render_template, request, jsonify, send_file, send_from_directory
import os


app = Flask(__name__)

fall_detection_result = False

BODY_PARTS = {"Nose": 0, "Neck": 1, "RShoulder": 2, "RElbow": 3, "RWrist": 4,
              "LShoulder": 5, "LElbow": 6, "LWrist": 7, "RHip": 8, "RKnee": 9,
              "RAnkle": 10, "LHip": 11, "LKnee": 12, "LAnkle": 13, "REye": 14,
              "LEye": 15, "REar": 16, "LEar": 17}

POSE_PAIRS = [["Neck", "RShoulder"], ["Neck", "LShoulder"], ["RShoulder", "RElbow"],
              ["RElbow", "RWrist"], ["LShoulder", "LElbow"], ["LElbow", "LWrist"],
              ["Neck", "RHip"], ["RHip", "RKnee"], ["RKnee", "RAnkle"], ["Neck", "LHip"],   
              ["LHip", "LKnee"], ["LKnee", "LAnkle"], ["Neck", "Nose"], ["Nose", "REye"],
              ["REye", "REar"], ["Nose", "LEye"], ["LEye", "LEar"]]

EDGES = {
    (BODY_PARTS["Nose"], BODY_PARTS["Neck"]): 'm',
    (BODY_PARTS["Neck"], BODY_PARTS["RShoulder"]): 'm',
    (BODY_PARTS["Neck"], BODY_PARTS["LShoulder"]): 'c',
    (BODY_PARTS["RShoulder"], BODY_PARTS["RElbow"]): 'm',
    (BODY_PARTS["RElbow"], BODY_PARTS["RWrist"]): 'm',
    (BODY_PARTS["LShoulder"], BODY_PARTS["LElbow"]): 'c',
    (BODY_PARTS["LElbow"], BODY_PARTS["LWrist"]): 'c',
    (BODY_PARTS["Neck"], BODY_PARTS["RHip"]): 'y',
    (BODY_PARTS["RHip"], BODY_PARTS["RKnee"]): 'm',
    (BODY_PARTS["RKnee"], BODY_PARTS["RAnkle"]): 'm',
    (BODY_PARTS["Neck"], BODY_PARTS["LHip"]): 'y',
    (BODY_PARTS["LHip"], BODY_PARTS["LKnee"]): 'c',
    (BODY_PARTS["LKnee"], BODY_PARTS["LAnkle"]): 'c',
    (BODY_PARTS["Nose"], BODY_PARTS["REye"]): 'm',
    (BODY_PARTS["REye"], BODY_PARTS["REar"]): 'm',
    (BODY_PARTS["Nose"], BODY_PARTS["LEye"]): 'c',
    (BODY_PARTS["LEye"], BODY_PARTS["LEar"]): 'c',
}
net = cv2.dnn.readNetFromTensorflow(r"./graph_opt.pb")

inWidth = 368
inHeight = 368
thr = 0.2

fall_model_path = "C:\Users\user\Downloads\send\export\fall-detection-model.h5"
fall_model = load_model(fall_model_path)

# Load the MoveNet interpreter
interpreter = tf.lite.Interpreter(model_path="3.tflite")
interpreter.allocate_tensors()

def run_movenet(image):
    # Run MoveNet model to detect keypoints
    input_details = interpreter.get_input_details()
    output_details = interpreter.get_output_details()

    # Preprocess the image
    input_image = cv2.resize(image, (192, 192))  # Resize image to match model input size
    input_image = np.expand_dims(input_image, axis=0)  # Add batch dimension
    input_image = tf.cast(input_image, dtype=tf.float32)

    # Set input tensor
    interpreter.set_tensor(input_details[0]['index'], input_image)
    interpreter.invoke()

    # Get the keypoints from the interpreter
    keypoints_with_scores = interpreter.get_tensor(output_details[0]['index'])
    
    # Extract only the (x, y) coordinates of each keypoint
    keypoints = keypoints_with_scores[0][:, :, :2]  # Extract (x, y) coordinates, ignore the confidence scores
    
    return keypoints

def draw_keypoints(frame, keypoints):
    y, x, c = frame.shape
    shaped = np.squeeze(np.multiply(keypoints, [[y, x]]))
    
    for kp in shaped:
        ky, kx = kp
        # Draw the keypoint
        cv2.circle(frame, (int(kx), int(ky)), 4, (0, 255, 0), -1) 
        # Display coordinates next to the keypoint
        text = f"({int(kx)}, {int(ky)})"
        cv2.putText(frame, text, (int(kx) + 10, int(ky) + 10), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255), 1)
            
def draw_connections(frame, keypoints, edges):
    y, x, c = frame.shape
    shaped = np.squeeze(np.multiply(keypoints, [[y, x]]))
    
    for edge, color in edges.items():
        p1, p2 = edge
        y1, x1 = shaped[p1]
        y2, x2 = shaped[p2]
        
        cv2.line(frame, (int(x1), int(y1)), (int(x2), int(y2)), (0, 0, 255), 2)

def pose_estimation(frame, args=None):
    frameWidth = frame.shape[1]
    frameHeight = frame.shape[0]

    net.setInput(cv2.dnn.blobFromImage(frame, 1.0, (inWidth, inHeight), (127.5, 127.5, 127.5), swapRB=True, crop=False))
    out = net.forward()
    out = out[:, :18, :, :]

    assert (len(BODY_PARTS) == out.shape[1])

    points = []
    confidences = []

    for i in range(len(BODY_PARTS)):
        heatMap = out[0, i, :, :]

        _, conf, _, point = cv2.minMaxLoc(heatMap)
        x = (frameWidth * point[0]) / out.shape[3]
        y = (frameHeight * point[1]) / out.shape[2]

        points.append((int(x), int(y)) if conf > thr else None)
        confidences.append(conf)

    for pair in POSE_PAIRS:
        partFrom = pair[0]
        partTo = pair[1]
        assert (partFrom in BODY_PARTS)
        assert (partTo in BODY_PARTS)

        idFrom = BODY_PARTS[partFrom]
        idTo = BODY_PARTS[partTo]

        if points[idFrom] and points[idTo]:
            cv2.line(frame, points[idFrom], points[idTo], (0, 255, 0), 3)
            cv2.ellipse(frame, points[idFrom], (3, 3), 0, 0, 360, (0, 0, 255), cv2.FILLED)
            cv2.ellipse(frame, points[idTo], (3, 3), 0, 0, 360, (0, 0, 255), cv2.FILLED)

    t, _ = net.getPerfProfile()
    freq = cv2.getTickFrequency() / 1000
    cv2.putText(frame, '%.2fms' % (t / freq), (10, 20), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 0, 0))

    return frame, points, confidences

def predict_image(image_path):
    # Load the image
    image = cv2.imread(image_path)
    
    keypoints = run_movenet(image)
    
    flattened_keypoints = keypoints.flatten().reshape(1, -1)
    prediction = fall_model.predict(flattened_keypoints)

    if prediction> 0.5:
      return True
    else:
      return False

@app.route('/')
def homepage():
    return send_file("./fall/trained_home.handlebars")

@app.route('/<path:path>')
# Serve index.html which contains the code of webcam in Javascript
def index(path):
    return send_from_directory('static', path)

@app.route('/upload', methods=['POST'])
def upload():
    try:
        global fall_detection_result
        image = request.files.get('content')
        image.save("in.png")
        img = cv2.imread("in.png")

        estimated_image, points, confidences = pose_estimation(img)

        # Predict fall using the image
        fall_detected = predict_image("in.png")
        fall_detection_result = fall_detected
        print(fall_detected)

        cv2.imwrite("out.png", estimated_image)

        return "ok"
    
    except Exception as e:
        print(e)

@app.route('/get_fall_detection_result', methods=['GET'])
def get_fall_detection_result():
    global fall_detection_result
    return str(fall_detection_result)

@app.route('/get', methods=['GET'])
def get():
    return send_file("./out.png")

if __name__ == '__main__':
     app.run(host='0.0.0.0', port=5000)